File: Problems/0100_0199/0124_Binary_Tree_Maximum_Path_Sum/Project_Go/solution/OperateTreeNode.go
package solution

import (
	"fmt"
	"math"
	"strconv"
	"strings"
)

func CreateTreeNode(flds string) *TreeNode {
	return createSubTreeNode(strings.Split(flds, ","), 0, 0)
}

func createSubTreeNode(flds []string, depth int, pos int) *TreeNode {
	if len(flds) == 0 {
		return nil
	}

	curPos := 0
	for i := 0; i < depth; i++ {
		curPos += int(math.Pow(2.0, float64(i)))
	}

	if curPos+pos > len(flds)-1 {
		return nil
	}

	if flds[curPos+pos] == "null" {
		return nil
	}

	val, err := strconv.Atoi(flds[curPos+pos])
	if err != nil {
		fmt.Printf("createTreeNode() Error ... flds[%d] = %s\n", curPos+pos, flds[curPos+pos])
		return nil
	}

	/*
		node := new(TreeNode)
		node.Val = val
		node.Left = createSubTreeNode(flds, depth+1, 2*pos)
		node.Right = createSubTreeNode(flds, depth+1, 2*pos+1)
	*/
	node := &TreeNode{val, createSubTreeNode(flds, depth+1, 2*pos), createSubTreeNode(flds, depth+1, 2*pos+1)}
	return node
}

var resultList []string

func TreeToStaircaseString(node *TreeNode) string {
	resultList = []string{}
	outStr := TreeToStaircaseSubString(node, 0)
	resultList = nil

	return outStr
}

func TreeToStaircaseSubString(node *TreeNode, n int) string {
	if node == nil {
		return ""
	}

	if len(resultList) <= n {
		resultList = append(resultList, "("+strconv.Itoa(node.Val)+")")
	} else {
		resultList[n] = resultList[n] + ",(" + strconv.Itoa(node.Val) + ")"
	}

	if node.Left != nil {
		TreeToStaircaseSubString(node.Left, n+1)
	}
	if node.Right != nil {
		TreeToStaircaseSubString(node.Right, n+1)
	}

	resultStr := ""
	for _, s := range resultList {
		resultStr += s + "\n"
	}

	return resultStr
}

func Tree2str(node *TreeNode) string {
	if node == nil {
		return ""
	}

	resultStr := strconv.Itoa(node.Val)

	if node.Left == nil && node.Right == nil {
		return resultStr
	}

	resultStr += "(" + Tree2str(node.Left) + ")"
	if node.Right != nil {
		resultStr += "(" + Tree2str(node.Right) + ")"
	}

	return resultStr
}

File: Problems/0500_0599/0563_Binary_Tree_Tilt/Project_Python3/TreeNode/OperateTreeNode.py
from typing import List, Dict, Tuple

from .TreeNode import TreeNode

class OperateTreeNode:
#   def createTreeNode(self, flds):
    def createTreeNode(self, flds: List[str], depth: int = 0, pos: int = 0) -> TreeNode:
        if pos >= len(flds) or len(flds) <= 0:
            return None
        cur_pos = 0
        for i in range(depth):
            cur_pos += 2 ** i
        if cur_pos + pos > len(flds) - 1:
            return None
        if flds[cur_pos + pos] == 'null':
            return None
        node = TreeNode(int(flds[cur_pos + pos]))
        node.left = self.createTreeNode(flds, depth + 1, 2*pos)
        node.right = self.createTreeNode(flds, depth + 1, 2*pos + 1)
        return node

    def treeToStaircaseString(self, node: TreeNode) -> str:
        self.resultList = []
        self.getStaircaseSubString(node, 0)
        return self.resultListToString()

    def getStaircaseSubString(self, node: TreeNode, n: int) -> None:
        if node is None:
            return
        if len(self.resultList) <= n:
            self.resultList.append("(" + str(node.val) + ")")
        else:
            self.resultList[n] += ",(" + str(node.val) + ")"
        if node.left is not None:
            self.getStaircaseSubString(node.left, n + 1)
        if node.right is not None:
            self.getStaircaseSubString(node.right, n + 1)
        return

    def resultListToString(self) -> str:
        resultStr = "\n".join(self.resultList)
        self.resultList.clear()
        return resultStr

    def tree2str(self, node: TreeNode) -> str:
        """
        :type node: TreeNode
        :rtype: str
        """
        if node is None:
            return ""

        resultStr = str(node.val)

        if node.left is None and node.right is None:
            return resultStr

        resultStr += "(" + self.tree2str(node.left) + ")"
        if node.right is not None:
            resultStr += "(" + self.tree2str(node.right) + ")"

        return resultStr
